### ISSUE TITLE
PartitionIdentity - design

### ISSUE THREAD

**Issue by rogeralsing** [AsynkronIT/protoactor-dotnet#822]
#821

The update of PartitionIdentity based on membership hash seems to work well.
There are however some old design issues with the PartitionId that could be improved upon.

1.
e.g. when an actor terminates, it sends an ActivationTerminated event to the node that we believe is the owner of the identity.
This is probably fine in most cases, but it doesn't seem robust to me.
it might end up sending it to the incorrect node if there is a topology change going on right now.

Maybe cluster broadcast activation terminations instead, so that any member that might hold on to this identity can act upon it?

2.
There is also a case when the DefaultClusterContext wants to tell the Identity Lookup that a PID no longer exists, this is not implemented at all in the partition identity. and thus, it will keep telling the consumer to use this incorrect pid.

Maybe the PartitionIdentityLookup should act on this call, and send a message to the identity owner saying "remove this pid" ?
Or, maybe this should be treated the same way as 1. ?

3.
We use Reenter for activation tasks, we however don't have any clean way to deal with failed activations.
we pick a node to activate on, call it and ask it to activate, then wait for reentrancy on that task.
The task may however come back successfully, non faulted, but with no pid. (unsure on why this is atm)
in such case, we do not have a retry in this activation flow.

This should probably be redesigned

**Comment by rogeralsing** [comment]
The current WiP PR now resolved 1 and 2 using broadcast.

**Comment by rogeralsing** [comment]
Currently debugging why a handful requests fails in the cluster.RequestAsync in ClusterBenchmark.
If we pass a long timeout to it, say 30 seconds, it will run for the full duration and then return null.

Requesting the same cluster identity again after now succeeds.

It looks like some form of eventual consistency going on. e.g. in the IdentityLookup.RemovePid.
The strange bit is that cluster.RequestAsync does retries and should purge any invalid pids. but it seems like we get stuck for the entire duration there in the rare cases where this happens

**Comment by rogeralsing** [comment]
Some more debugging, it turns out that I was wrong, when it happens, it happens for the same identity over and over.

<img width="1440" alt="Skärmavbild 2021-03-29 kl  07 57 52" src="https://user-images.githubusercontent.com/647031/112792862-85273280-9064-11eb-8e69-8e0c1ed08430.png">

**Comment by rogeralsing** [comment]
Debugging branch here https://github.com/AsynkronIT/protoactor-dotnet/pull/838

**Comment by rogeralsing** [comment]
GetAsync in PartitionIdentityLookup used the request cancellation token. e.g. if that is 20 sec, and we use that here, getting the PID will hang for 20 sec.

Replaced this with new ct's

This shows that we are ending up on ct timeout on the PartitionIdentityLookup.
Question is then, why don't the remote node respond?

**Comment by rogeralsing** [comment]
<img width="1319" alt="Skärmavbild 2021-03-29 kl  08 41 56" src="https://user-images.githubusercontent.com/647031/112796711-a4c15980-906a-11eb-83ad-2830541ebe74.png">

It still requests the identity from the node that just left.
We see the topology event coming in after this: CT1682241162
This looks correct as it was running non graceful shutdowns and we only know of the topology change once we get the event

**Comment by rogeralsing** [comment]
TLDR;

It is important that the RequestAsync timeout is longer than the timeouts used to resolve the pids.
This goes for all identity providers. if not, you will get a null response due to timeouts.

if the request timeout is 20 seconds, and the timeout to resolve a pid is 5 seconds.
In a worst-case scenario, we will have at least ~4 pid resolves during this 20 sec period.

**Comment by rogeralsing** [comment]
We are still seeing occasional dead virtual actors, and from the logs we see it's the same actors that keeping being dead.

<img width="1312" alt="Skärmavbild 2021-03-29 kl  09 18 32" src="https://user-images.githubusercontent.com/647031/112800302-c1ac5b80-906f-11eb-9975-e2e33f62ec15.png">
<img width="862" alt="Skärmavbild 2021-03-29 kl  09 18 25" src="https://user-images.githubusercontent.com/647031/112800307-c2dd8880-906f-11eb-850c-978e51fb51c1.png">

It does resolve itself after a topology change.

**Comment by rogeralsing** [comment]
Found, patched, and pushed.
